### PR TITLE
Fifth assessment Joel

### PR DESCRIPTION
**Pull request by devJoelMichael**
Final-completed is the one that has all the answers :D

### CODE REVIEW

File: code-challenges.js [hunk 1]
```diff
@@ -2,29 +2,101 @@
...
 
+// ############ CONDENSED VERSION ####################
 
+const cssrz = (str) => {
+  console.log("What is str:", str)
+  return str.replace(/\a/g, '4')
+    .replace(/\e/g, '3')
+    .replace(/\i/g, '1')
+    .replace(/\o/g, '0')
+}
```
Comment (rachaelstanislaw) [inline code-challenges.js:46]: nice work on both of these solutions! very organized code and easy to follow

File: code-challenges.js [hunk 2]
```diff
@@ -2,29 +2,101 @@
...
+console.log("FIRST: \n")
+
 var secretCodeWord1 = "lackadaisical"
 // Expected output: "l4ck4d41s1c4l"
 var secretCodeWord2 = "gobbledygook"
 // Expected output: "g0bbl3dyg00k"
 
+// ########################### ORIGINAL ##################################
+
+const cssr = (str) => {
```
Comment (rachaelstanislaw) [inline code-challenges.js:14]: not sure what the name of this function means?

File: code-challenges.js [hunk 3]
```diff
@@ -2,29 +2,101 @@
...
 
 
 // --------------------2) Create a function that takes in an array and returns all the words that contain the letter a.
+console.log("SECOND: \n")
 
 var arrayOfWords = ["Apple", "Banana", "Plum", "Cherry", "Kiwi", "Peach"]
 // Expected output: "Apple" "Banana" "Peach"
 
+// first create a function that will take in a string
+const sweetAzz = (array) => {
```
Comment (rachaelstanislaw) [inline code-challenges.js:59]: not sure what this function name means either?

File: code-challenges.js [hunk 4]
```diff
@@ -2,29 +2,101 @@
...
+  // use a filter to sort through the array
+  return array.filter(word => {
+    // console.log("1:, \n", word)
+    word.split("")
+    if (word.toLowerCase().includes("a") || word.toUpperCase().includes("A")) {
```
Comment (rachaelstanislaw) [inline code-challenges.js:64]: you dont need the `or` statement here. by making the whole word lowercase and making the comparison, you can assume all uppercase letters are now lowercase.

File: code-challenges.js [hunk 5]
```diff
@@ -2,29 +2,101 @@
...
 
 var arrayOfWords = ["Apple", "Banana", "Plum", "Cherry", "Kiwi", "Peach"]
 // Expected output: "Apple" "Banana" "Peach"
 
+// first create a function that will take in a string
+const sweetAzz = (array) => {
+  // use a filter to sort through the array
+  return array.filter(word => {
+    // console.log("1:, \n", word)
+    word.split("")
```
Comment (rachaelstanislaw) [inline code-challenges.js:63]: what is the purpose of splitting the word?

File: code-challenges.js [hunk 6]
```diff
@@ -2,29 +2,101 @@
...
+    // console.log(cards)
+    if (cards[0] === cards[1] && cards[1] === cards[2] && cards.length >= 3
+      && cards[3] === cards[4]) {
+      // console.log(cards)
+      return "true"
+    } else {
+      return "false"
+    }
+  }
+}
```
Comment (rachaelstanislaw) [inline code-challenges.js:97]: the condition to be met in the loop will only work with these exact test variables. it doesn't work when the cards are out of order.

File: code-challenges.js [hunk 7]
```diff
@@ -2,29 +2,101 @@
...
 
 var hand1 = [5, 5, 5, 3, 3]
 // Expected output: true
 var hand2 = [5, 5, 3, 3, 4]
 // Expected output: false
 var hand3 = [5, 5, 5, 5, 4]
 // Expected output: false
+
+const fullHouse = (cards) => {
+  for (let i = 0; i < cards.length; i++) {
```
Comment (rachaelstanislaw) [inline code-challenges.js:87]: what is the purpose of the loop? if you take it out, you get the same outcome

File: interview-questions.md [hunk 8]
```diff
@@ -7,65 +7,70 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-  Researched answer:
+  Researched answer: Migrations are a convenient way to alter your database schema over time in a consistent way. You can think of each migration as a "new version" of the database.
 
 
 
 4. Describe as many possible relationships between a model called Person and a model called Computer as you can think of. Based on each description, which model would hold the foreign key?
 
-  Your answer:
+  Your answer: rails generate model Person/Computer (i know thats not how you would write it) brain_power:string basic_functions:string spill_water_on_us_we_hate_it:boolean * I could go on and on but to answer the question I believe basic_functions or brain_power would hold the foreign key.
```
Comment (rachaelstanislaw) [inline interview-questions.md:34]: this is more asking about relationships between a Person model and Computer model. AKA: Person has_many computers, computer belongs_to person, etc. in this case, Computer (or belongs_to side) would house the foreign key

File: interview-questions.md [hunk 9]
```diff
@@ -7,65 +7,70 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-  Researched answer:
+  Researched answer: * I will be honest I don't know how I am supposed to research this *
```
Comment (rachaelstanislaw) [inline interview-questions.md:36]: I would google "model relationships in sql" which might have helped with understanding the question

File: interview-questions.md [hunk 10]
```diff
@@ -7,65 +7,70 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
+- RESTful routes - REST is an architectural style that is used in Rails by default. RESTful design helps to create default routes (by using resources keyword) in the config/routes.rb file.
+
+- API - An application programming interface is a computing interface which defines interactions between multiple software intermediaries. It defines the kinds of calls or requests that can be made, how to make them, the data formats that should be used, the conventions to follow, etc.
+
+- Endpoints - An API endpoint is a point at which an application program interface (API) -- the code that allows two software programs to communicate with each other -- connects with the software program. APIs work by sending requests for information from a web application or web server and receiving a response.
+
+- Strong params - It provides an interface for protecting attributes from end-user assignment. This makes Action Controller parameters forbidden to be used in Active Model mass assignment until they have been explicitly enumerated.
+
+In addition, parameters can be marked as required and flow through a predefined raise/rescue flow to end up as a 400 Bad Request with no effort.
```
Comment (rachaelstanislaw) [inline interview-questions.md:76]: good research!

File: rails-questions.md [hunk 11]
```diff
@@ -2,33 +2,57 @@
 
 1. What is the command to create a new Rails application with a Postgres database?
 
-
+  * rails new i_feel_confident_in_this_one -d postgresql -T and then we must cd into i_feel_confident_in_this_one and then put rails db:create and open that bad boy up and have some fun!
```
Comment (rachaelstanislaw) [inline rails-questions.md:5]: 👍🏻

File: rails-questions.md [hunk 12]
```diff
@@ -2,33 +2,57 @@
...
 
 2. Explain each section of the following route:  get '/library/:id' => 'book#show'
 
-get -
-library -
-:id -
-book -
-show -
+get - get is pretty much what it says. It is getting the key/value/etc from a specific place in your database (through the route).
```
Comment (rachaelstanislaw) [inline rails-questions.md:9]: HTTP verb

File: rails-questions.md [hunk 13]
```diff
@@ -2,33 +2,57 @@
...
+
+library - I believe that library is the class. Like this?
```
Comment (rachaelstanislaw) [inline rails-questions.md:11]: URL

File: rails-questions.md [hunk 14]
```diff
@@ -2,33 +2,57 @@
...
+  class Library < ApplicationRecord
+    has_many :books
+  end
+
+:id - this is the key or value (I know they are different I think this one is the value? I have just heard tutorials call them different things hahaha. (I might be getting my terminology wrong so im sorry, I do need to memorize some of those cuz that will mess me up.)
```
Comment (rachaelstanislaw) [inline rails-questions.md:16]: param

File: rails-questions.md [hunk 15]
```diff
@@ -2,33 +2,57 @@
...
+
+book - this would be the "MainController" for the examples we used last. I am trying to remember the actual term. I believe it would be whatever you do like this?
```
Comment (rachaelstanislaw) [inline rails-questions.md:18]: book is the name of the controller

File: rails-questions.md [hunk 16]
```diff
@@ -2,33 +2,57 @@
...
+  class Library < ApplicationRecord
+    has_many :books
+  end
+
+:id - this is the key or value (I know they are different I think this one is the value? I have just heard tutorials call them different things hahaha. (I might be getting my terminology wrong so im sorry, I do need to memorize some of those cuz that will mess me up.)
+
+book - this would be the "MainController" for the examples we used last. I am trying to remember the actual term. I believe it would be whatever you do like this? 
+rails g controller book ? I think hahaha
+
+show - I believe this is the method name we are using? So like for this one it would be:
```
Comment (rachaelstanislaw) [inline rails-questions.md:21]: 👍🏻 show is the method in the controller

File: rails-questions.md [hunk 17]
```diff
@@ -2,33 +2,57 @@
...
+  def show
+    @show (then whatever would complete what its trying to do) 
+  end
 
 
 3. Oops, I forgot to add a foreign key to my model. Describe the steps to remedy this mistake.
 
+  * You would want to use CREATE INDEX to index them and then you would want to use ADD CONSTRAINT statement to add the FOREIGN KEY constraint to the columns.
```
Comment (rachaelstanislaw) [inline rails-questions.md:29]: look at the section in the syllabus and watch Andee's lecture on rails migrations for this one

File: rails-questions.md [hunk 18]
```diff
@@ -2,33 +2,57 @@
...
+  validates :pants, presence: true 
+  validates :shoes, presence: true 
+
 
 
 
 
 6. What is the naming convention for generating a Rails model? What is the naming convention for the table that is generated with this command?
+
+  * rails generate model Pokemon name:string type:string height:string weight:integer date_caught:dateTime
```
Comment (rachaelstanislaw) [inline rails-questions.md:56]: the naming convention for a model is PascalCase and the naming convention of a table is snake_case :)

**Review by rachaelstanislaw — APPROVED** [review]
your code challenges were really organized and easy to follow so great work with that. I might look into iteration and built-in methods a little more since there were some parts of your code that were a little redundant or unnecessary. nothing was broken in the code, but it could make your life a little easier and will help avoid a bunch of extra key strokes :) also, some of the interview questions were potentially misunderstood. we are always happy to clarify questions if you have them :) otherwise good work! only one more assessment to go! woo!

File: code-challenges.js (hunk 2 above)
Comment (devJoelMichael) [reply code-challenges.js:14]: hahaha I was trying to be goofy. I was thinking CSSR, like russian spys and what not, but I could have made it KGB. Also in real life and work I probably shouldn't make just weird names lol

File: code-challenges.js (hunk 3 above)
Comment (devJoelMichael) [reply code-challenges.js:59]: LOL I forgot I left this in here, its supposed to be sweet As, like just find those sweet As....but I was feeling frusterated at the time and did that and then forgot to change it....sooo hahaha sorry

File: code-challenges.js (hunk 4 above)
Comment (devJoelMichael) [reply code-challenges.js:64]: Ahhhhhh ok, that makes sense.

**Pull request merged**